File: typescript/linked-list/linked-list.ts
class Node<TElement> {
  _data: TElement
  _next: Node<TElement> | null

  constructor(element: TElement) {
    this._data = element
    this._next = null
  }
}

export class LinkedList<TElement> {
  _head: Node<TElement> | null = null

  public push(element: TElement): void {
    const node = new Node(element)

    if(!this._head) {
      this._head = node
    } else {
      let current = this._head
      while(current._next) {
        current = current._next
      }
      current._next = node
    }
  }

  public pop(): TElement | null {
    let current = this._head

    if(current?._next) {
      let next = current._next
      while(next._next) {
        current = next
        next = next._next
      }
      current._next = null
      return next._data
    } else if(current) {
      this._head = null
      return current._data
    } else {
      return null
    }
  }

  public shift(): TElement | null {
    const current = this._head

    if(current) {
      this._head = current._next
      return current._data
    } else {
      return null
    }
  }

  public unshift(element: TElement): void {
    const node = new Node(element)
    const current = this._head

    this._head = node
    node._next = current
  }

  public delete(element: TElement): void {
    let current = this._head
    let next = current?._next
    
    if(current?._data === element) {
      this._head = current._next
      return
    }

    while(current && next) {
      if(next._data === element) {
        current._next = next._next
      }
      current = next
      next = next._next
    }
  }

  public count(): number {
    let current = this._head
    let count = current ? 1 : 0
    while(current?._next) {
      current = current._next
      count++
    }
    return count
  }
}

File: ruby/linked-list/linked_list.rb
require 'pry'

class Deque
  attr_accessor :head, :tail

  def push(elem)
    node = Node.new(elem)

    if head && tail
      tail.next_node = node
      node.prev_node = tail
      @tail = node
    else
      @head = @tail = node
    end
  end

  def pop
    node = tail
    if head == tail
      @head = nil
    else
      @tail = tail.prev_node
      tail.next_node = nil
    end
    node.value
  end

  def shift
    node = head
    if head == tail
      @head = nil
    else
      @head = head.next_node
      head.prev_node = nil
    end
    node.value
  end

  def unshift(elem)
    node = Node.new(elem)

    if head && tail
      head.prev_node = node
      node.next_node = head
      @head = node
    else
      @head = @tail = node
    end
  end
end

class Node
  attr_accessor :value, :next_node, :prev_node

  def initialize(value)
    @value = value
  end
end
